src/ConsumeOwnEventsService.cpp

```diff
@@ -0,0 +1,147 @@
+// Copyright (C) Martin Da Costa 2023 (martindc.merg@gmail.com)
+// This file is part of VLCB-Arduino project on https://github.com/SvenRosvall/VLCB-Arduino
+// Licensed under the Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License.
+// The full licence can be found at: http://creativecommons.org/licenses/by-nc-sa/4.0/
+
+#include <Streaming.h>
+#include "ConsumeOwnEventsService.h"
+#include <vlcbdefs.hpp>
+
+namespace VLCB
+{
+
+void ConsumeOwnEventsService::setController(Controller *cntrl)
+{
+  this->controller = cntrl;
+  this->module_config = cntrl->getModuleConfig();
+}
+
+/// if buffer has one or more stored items
+
+bool ConsumeOwnEventsService::available(void)
+{
+  if (size > 0)
+  {
+    return true;
+  }
+  else
+  {
+  return false;
+  }
+
+ // return (size > 0);
+}
+
+/// store an item to the buffer - overwrite oldest item if buffer is full
+/// never called from an interrupt context so we don't need to worry about interrupts
+
+void ConsumeOwnEventsService::put(const VlcbMessage * msg)
+{
+//  VlcbMessage msg;
+  memcpy((VlcbMessage*)&buffer[head], (const VlcbMessage *)msg, sizeof(VlcbMessage));
+
+  // if the buffer is full, this put will overwrite the oldest item
+
+  if (full)
+  {
+    tail = (tail + 1) % capacity;
+    ++numOverflows;
+  }
+
+  head = (head + 1) % capacity;
+  full = head == tail;
+  size = bufUse();
+  hwm = (size > hwm) ? size : hwm;
+  ++numPuts;        // Counts how many events put to buffer.
+  // DEBUG_SERIAL << ">COE Puts = " << numPuts << " Size = " << size <<endl;
+
+  return;
+}
+
+/// retrieve the next item from the buffer
+
+VlcbMessage *ConsumeOwnEventsService::get(void)
+{
+  VlcbMessage *p = nullptr;
+
+  if (size > 0)
+  {
+    p = &buffer[tail];
+    full = false;
+    tail = (tail + 1) % capacity;
+    size = bufUse();
+    ++numGets;        // Counts how many events got from buffer.
+    // DEBUG_SERIAL << ">COE Gets = " << numGets << endl;
+  }
+
+  return p;
+}
+
+/// peek at the next item in the buffer without removing it
+
+VlcbMessage *ConsumeOwnEventsService::peek(void)
+{
+  // should always call ::available first to avoid this
+  if (size == 0)
+  {
+    return nullptr;
+  }
+
+  return (&buffer[tail]);
+}
+
+/// clear all items
+
+void ConsumeOwnEventsService::clear(void)
+{
+  head = 0;
+  tail = 0;
+  full = false;
+  size = 0;
+
+  return;
+}
+
+/// recalculate number of items in the buffer
+
+byte ConsumeOwnEventsService::bufUse(void)
+{
+  size = capacity;
```

Comment:
Why are you assigning the object field here, and then assign the return value to "size". I would return the expression you set "size" to here. And if(full) return capacity.
Or rename the function to updateSize() and don't return anything.

Reply:
Actually, this whole thing is convoluted in a most unnecessary way. If the buffer is full, it equals capacity, which is numItems.  If it's not full, then that is the only time we need to do anything.  Let me think about this one.

Reply:
I am ok with the logic. What I am commenting on is that you both set the field "size" and return a value which is then used to assign to "size". Duplicate work.

Reply:
Yes, I agree that it does duplicate and become overcomplicated.  I'll have a look at it later this morning in hospital before the chemicals get to me!